### PR TITLE
MES-1775 - Added postal address to comms page

### PR DESCRIPTION
**Pull request by roberthall12**
## Description and relevant Jira numbers

Added the postal address to the communication capture page. Also includes validation fixes for the page.
https://jira.i-env.net/browse/MES-1775

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [x] commits are squashed

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github
- [x] Any new reusable component/widget added to component library on Confluence

- [x] Screenshot(s) captured on the physical device (if applicable)
- [x] Tested by QA
- [x] PO's approval
Screen Shot 2019-05-20 at 11 15 45

### CODE REVIEW

File: src/modules/tests/candidate/__tests__/candidate-selector.spec.ts [hunk 1]
```diff
@@ -41,4 +51,26 @@ describe('candidate selector', () => {
       expect(formatDriverNumber('ABCDE123456Z78YX')).toBe('ABCDE 123456 Z78YX');
     });
   });
+
+  describe('getPostalAddress', () => {
+    it('should output the address', () => {
+      expect(getPostalAddress(candidate).addressLine1).toEqual('1 Example Street');
```
Comment (johnmorleybjss) [inline src/modules/tests/candidate/__tests__/candidate-selector.spec.ts]: Could this test be simplified to comparing a higher level reference?
```
expect(getPostalAddress(candidate)).toEqual(candidate.candidateAddress);
```

File: src/modules/tests/candidate/candidate.selector.ts [hunk 2]
```diff
@@ -20,3 +20,12 @@ export const formatDriverNumber = (driverNumber: string) => {
 };
 
 export const getCandidateEmailAddress = (candidate: Candidate) => candidate.emailAddress ? candidate.emailAddress : '';
+
+export const getPostalAddress = (candidate: Candidate) => candidate.candidateAddress;
+
+export const formatAddress = (address: Address) => {
+  const regex = new RegExp('[0-9]', 'g');
```
Comment (Cavalier07) [inline src/modules/tests/candidate/candidate.selector.ts]: this shouldn't be in the selector as it isn't selecting any data from the state

File: src/pages/communication-page/components/postal-address/postal-address.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  formGroup: FormGroup;
+
+  @Input()
+  postalAddress: Address;
+
+  @Input()
+  isPostalAddressChosen: boolean;
+
+  @Output()
+  postalRadioSelect = new EventEmitter();
```
Comment (johnmorleybjss) [inline src/pages/communication-page/components/postal-address/postal-address.ts]: Type parameter would be nice here.

File: src/pages/communication-page/__tests__/communication.spec.ts [hunk 4]
```diff
@@ -109,6 +116,17 @@ describe('CommunicationPage', () => {
       expect(component).toBeDefined();
     });
 
+    describe('DOM', () => {
+      it('should display the correct address fields', () => {
```
Comment (Cavalier07) [inline src/pages/communication-page/__tests__/communication.spec.ts]: To make sure every possibility is covered, splitting this into 2 tests one which tests everything being there and one which tests everything missing.

File: src/pages/communication-page/components/postal-address/postal-address.scss [hunk 5]
```diff
@@ -0,0 +1,8 @@
+.displayAddress {
```
Comment (johnmorleybjss) [inline src/pages/communication-page/components/postal-address/postal-address.scss]: I think it's nice to use the convention that we use in the `.scss` for other components that the top-level selector is the component directive selector.

Comment (roberthall12) [reply src/pages/communication-page/components/postal-address/postal-address.scss]: fixed

File: src/pages/communication-page/components/postal-address/postal-address.ts (hunk 3 above)
Comment (roberthall12) [reply src/pages/communication-page/components/postal-address/postal-address.ts]: fixed

File: src/modules/tests/candidate/__tests__/candidate-selector.spec.ts (hunk 1 above)
Comment (roberthall12) [reply src/modules/tests/candidate/__tests__/candidate-selector.spec.ts]: fixed

File: src/pages/communication-page/__tests__/communication.spec.ts (hunk 4 above)
Comment (roberthall12) [reply src/pages/communication-page/__tests__/communication.spec.ts]: fixed

File: src/modules/tests/candidate/candidate.selector.ts (hunk 2 above)
Comment (roberthall12) [reply src/modules/tests/candidate/candidate.selector.ts]: fixed

File: src/pages/communication-page/__tests__/communication.spec.ts [hunk 6]
```diff
@@ -109,6 +112,40 @@ describe('CommunicationPage', () => {
...
+      it('should display the no address fields when address is empty', () => {
+        expect(fixture.debugElement.query(By.css('#addressLine1'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#addressLine2'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#addressLine3'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#addressLine4'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#addressLine5'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#postcode'))).toBeNull();
+      });
+
+      mockCandidate.candidateAddress = address;
```
Comment (Cavalier07) [inline src/pages/communication-page/__tests__/communication.spec.ts]: this shouldn't be here

File: src/pages/communication-page/communication.ts [hunk 7]
```diff
@@ -204,4 +219,11 @@ export class CommunicationPage extends BasePageComponent {
   verifyCandidateChoseProvidedEmail() {
     return (this.communicationEmail === '') || (this.communicationEmail === this.candidateProvidedEmail);
   }
+
+  formatAddress (address: Address): Address {
```
Comment (Cavalier07) [inline src/pages/communication-page/communication.ts]: You need some tests for this

File: src/pages/communication-page/components/postal-address/postal-address.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Component, Input, Output, EventEmitter } from '@angular/core';
+import { Address } from '@dvsa/mes-test-schema/categories/B';
+import { FormGroup } from '@angular/forms';
+
+@Component({
+  selector: 'postal-address',
+  templateUrl: 'postal-address.html',
+})
+export class PostalAddressComponent {
```
Comment (Cavalier07) [inline src/pages/communication-page/components/postal-address/postal-address.ts]: You have no tests for this componenet

File: src/pages/communication-page/__tests__/communication.spec.ts (hunk 6 above)
Comment (roberthall12) [reply src/pages/communication-page/__tests__/communication.spec.ts]: moved

File: src/pages/communication-page/__tests__/communication.spec.ts [hunk 9]
```diff
@@ -117,14 +113,37 @@ describe('CommunicationPage', () => {
...
+        expect(fixture.debugElement.query(By.css('#addressLine3'))).toBeNull();
         expect(fixture.debugElement.query(By.css('#addressLine4'))).toBeNull();
         expect(fixture.debugElement.query(By.css('#addressLine5'))).toBeNull();
+        expect(fixture.debugElement.query(By.css('#postcode'))).toBeNull();
+      });
+
+      mockCandidate.candidateAddress = address;
+
+      it('should display the correct address fields when address populated', () => {
+        expect(fixture.debugElement.query(By.css('#addressLine1'))).toBeDefined();
```
Comment (johnmorleybjss) [inline src/pages/communication-page/__tests__/communication.spec.ts]: Checking for the result of fixture query `toBeDefined` is a misleading test, as failure to find the element will yield `null`.

File: src/pages/communication-page/components/postal-address/postal-address.ts [hunk 10]
```diff
@@ -20,9 +20,9 @@ export class PostalAddressComponent {
   isPostalAddressChosen: boolean;
 
   @Output()
-  postalRadioSelect = new EventEmitter();
+  postalRadioSelect = new EventEmitter<boolean>();
```
Comment (johnmorleybjss) [inline src/pages/communication-page/components/postal-address/postal-address.ts]: You can do `EventEmitter<void>` if it's only ever going to be `true`.

**Pull request merged**